### PR TITLE
Fix multipackage

### LINKED ISSUE #1527: multipackage (MERGE) boken in PyInstaller 3.0
In PyInstaller 3.0 no much attention was paid to 'multipackage' feature and it is broken:
- MERGE should be fixed
- MERGE tests should be migrated to pytest

### PR DESCRIPTION
**Pull request by coreydexter**
Fix up the multipackage so that exe's can shared dependencies. Closes #1527

There are a few changes that needed to be done:

- Move the multipackage tests from old_suite to the new py.test suite.
- MERGE class: Correct the relative paths which reference shared dependencies from one exe to another 
- Bootloader: Fix a bug in the pyi_copy_file where when files were copied, an extra bit of data was been added if the file didn't finish perfectly on the buffer size. fwrite writes the entire buffer, not just up until a null terminator character, so this meant the last fwrite would include data from the previous fwrite which corrupted files.
- Bootloader: Fix a bug where an archive sometimes wouldn't be opened because malloc doesn't NULL values on initialisation.

### CODE REVIEW

**Comment by coreydexter** [comment]
@htgoebel any chance you could take a look at this?

**Comment by papr** [comment]
Ping for additional attention. It would be great to get this feature working again!

**Comment by ErikBjare** [comment]
@htgoebel Is there anything holding this back? I'd love to know if there's any chance we will get this back soon.

File: PyInstaller/building/api.py [hunk 1]
```diff
@@ -798,7 +799,11 @@ def _set_dependencies(self, analysis, path):
                 else:
                     dep_path = self._get_relative_path(path, self._dependencies[tpl[1]])
                     logger.debug("Referencing %s to be a dependecy for %s, located in %s" % (tpl[1], path, dep_path))
-                    analysis.dependencies.append((":".join((dep_path, tpl[0])), tpl[1], "DEPENDENCY"))
+                    analysis.dependencies.append((
+                        ":".join((dep_path, os.path.basename(tpl[1]))),
```
Comment (htgoebel) [inline PyInstaller/building/api.py]: How is this related to "Normalise path casing"?

File: bootloader/src/pyi_launch.c [hunk 2]
```diff
@@ -163,6 +163,14 @@ _get_archive(ARCHIVE_STATUS *archive_pool[], const char *path)
...
     archive->has_temp_directory = archive_pool[SELF]->has_temp_directory;
 
+    /*
+     * NULL the fp pointer as malloc does not initialise values
+     * which can result in a non-NULL value and will result in
+     * the archive not being opened and bogus data remaining
+     * in archive->fp
+     */
+    archive->fp = NULL;
+
```
Comment (htgoebel) [inline bootloader/src/pyi_launch.c]: Good catch. 

Maybe shorten the comment to something like `// maloc does not initialism values, ensure fp is NULL` and move the comment into the commit message? WDYT?

File: bootloader/src/pyi_utils.c [hunk 3]
```diff
@@ -607,15 +608,16 @@ pyi_copy_file(const char *src, const char *dst, const char *filename)
...
+        if (read_count <= 0 ) {
             if (ferror(in)) {
                 clearerr(in);
                 error = -1;
                 break;
             }
         }
         else {
-            int rc = fwrite(buf, 4096, 1, out);
+            size_t rc = fwrite(buf, 1, read_count, out);
```
Comment (htgoebel) [inline bootloader/src/pyi_utils.c]: Good catch.

File: news/1527.bugfix.rst [hunk 4]
```diff
@@ -0,0 +1,4 @@
+Fix multipackaging
+
+- Correct relative paths between shared dependencies
+- Fix issue when copying dependency files in the bootloader.
```
Comment (htgoebel) [inline news/1527.bugfix.rst]: Several remarks on this :-)

1. This rst-markup might not work as intented due to limitations og towncrier.

2. 1527 was the original issue, and 4303 is the pull-request. Thus there shall be several news fragments:
   - 1527.bugfix: about fixing multiprocessing (I suggest mentioning `MERGE` here)
   - 4303.bugfix: copy of that file
   - 4303.bootloader: about the bootloader issue (which issue? please elaborate)

3. Need a newsfragment for the 4th commit? Just name it e.g `fpnull.bootloader.rst`

File: PyInstaller/building/api.py [hunk 5]
```diff
@@ -780,10 +780,11 @@ def _merge_dependencies(self, args):
...
         """
         for analysis, _, _ in args:
-            path = os.path.abspath(analysis.scripts[-1][1]).replace(self._common_prefix, "", 1)
+            path = os.path.normcase(os.path.abspath(analysis.scripts[-1][1]))
+            path = path.replace(self._common_prefix, "", 1)
             path = os.path.splitext(path)[0]
-            if os.path.normcase(path) in self._id_to_path:
-                path = self._id_to_path[os.path.normcase(path)]
+            if path in self._id_to_path:
+                path = self._id_to_path[path]
```
Comment (htgoebel) [inline PyInstaller/building/api.py]: Is this actually needed? Please state in the commit message, why. If it is just for code clean up, I'd omit this change,

File: tests/functional/test_basic.py [hunk 6]
```diff
@@ -581,3 +581,23 @@ def test_several_scripts2(pyi_builder_spec):
...
+def test_spec_with_multipackage3(pyi_builder_spec):
+    pyi_builder_spec.test_spec('test_multipackage3.spec')
+
+
+def test_spec_with_multipackage4(pyi_builder_spec):
+    pyi_builder_spec.test_spec('test_multipackage4.spec')
+
+
+def test_spec_with_multipackage5(pyi_builder_spec):
+    pyi_builder_spec.test_spec('test_multipackage5.spec')
```
[the remark this answers is not in the document]
Comment (coreydexter) [reply tests/functional/test_basic.py]: Yeah sure that seems reasonable. I've updated these

File: tests/functional/conftest.py [hunk 7]
```diff
@@ -267,7 +267,8 @@ def _test_executables(self, name, args, runtime, run_from_path):
         assert exes != [], 'No executable file was found.'
         for exe in exes:
             # Try to find .toc log file. .toc log file has the same basename as exe file.
-            toc_log = os.path.join(_LOGS_DIR, os.path.basename(exe) + '.toc')
+            no_ext_base_exe_name = os.path.splitext(os.path.basename(exe))[0]
+            toc_log = os.path.join(_LOGS_DIR, no_ext_base_exe_name + '.toc')
```
[the remark this answers is not in the document]
Comment (coreydexter) [reply tests/functional/conftest.py]: No worries, done.

File: PyInstaller/building/api.py (hunk 1 above)
Comment (coreydexter) [reply PyInstaller/building/api.py]: Good question, I don't recall anymore - tpl[0] is the basename of tpl[1] so I'm not sure what the intention was here I've removed this and the tests still work.

File: bootloader/src/pyi_launch.c (hunk 2 above)
Comment (coreydexter) [reply bootloader/src/pyi_launch.c]: Done

File: PyInstaller/building/api.py (hunk 5 above)
Comment (coreydexter) [reply PyInstaller/building/api.py]: Not needed, was just removing the redundancy of the extra normcase. I have removed this change

File: news/1527.bugfix.rst (hunk 4 above)
Comment (coreydexter) [reply news/1527.bugfix.rst]: Done. I'm not 100% sure on the wording

File: PyInstaller/building/api.py (hunk 1 above)
Comment (coreydexter) [reply PyInstaller/building/api.py]: I recalled it was required, this is due to tpl[0] sometimes having the incorrect casing of the file name, which on operating system which have case-sensitive filepaths (eg Linux) results in the file not being loaded as the path is incorrect. I've added this comment to the comment message.

**Comment by sfgeorge** [comment]
@htgoebel Just checking, can you re-review this when you get a chance?  Thanks so much!

**Comment by MBjoern** [comment]
What's the status here? This bug is open since 4.5 years now...

**Comment by kmcquade** [comment]
@htgoebel, any chance you can re-review, as requested by @sfgeorge above? The parent issue for this (#1527) has been open for almost 5 years now :)

**Comment by Legorooj** [comment]
@htgoebel

**Comment by htgoebel** [comment]
@coreydexter I intentionally removed these "noqa: F401" changes :-)

**Comment by htgoebel** [comment]
@coreydexter I should have reasoned this: You added these comments to make the lint tests pass, wich basically is okay. But this is only test-code. And I count the information "renamed unchanged" more important than these errors. (Okay, what you most propably do not know: we only lint *changed* lines. So unless these lines will be touched again somewhen, lint will pass.)

**Comment by htgoebel** [comment]
@coreydexter So many thanks for solving this long-standing issue!

**Comment by Legorooj** [comment]
@htgoebel the docs need fixing. They still say it's broken.

**Comment by simaoafonso-pwt** [comment]
> @htgoebel the docs need fixing. They still say it's broken.

Besides this, is there any way to automatically merge several spec files?

**Comment by Legorooj** [comment]
Please open an issue if you have questions related to multipackage.
